Add sequence parallel strategy support.

GhostScreaming opened the pull request:
Still under development.

GhostScreaming committed:
[WIP] Add sequence parallel strategy support.

sneaxiy commented on fleetx/models/gpt_model/modeling_hybrid.py:
```diff
@@ -298,8 +357,7 @@ def forward(self,
         if self.use_recompute and self.recompute_granularity == "core_attn":
-            out, weights = recompute(
-                self.core_attn, q, k, v, attn_mask=attn_mask)
+            out, weights = recompute(self.core_attn, q, k, v)
```
Miss `attn_mask`?

sneaxiy commented on fleetx/models/gpt_model/modeling_hybrid.py:
```diff
@@ -325,16 +383,26 @@ def __init__(self,
+                    group = hcg.get_model_parallel_group()
+                    group.process_group.allreduce(grad).wait()
+                    return grad
+                self.norm.bias.register_hook(bias_all_reduce)
```
Missing `self.norm.weight.regsiter_hook(bias_all_reduce)`?

sneaxiy commented on fleetx/models/gpt_model/modeling_hybrid.py:
```diff
@@ -442,26 +512,38 @@ def __init__(self,
+                group = hcg.get_model_parallel_group()
+                group.process_group.allreduce(grad).wait()
+                return grad
+            self.norm1.bias.register_hook(bias_all_reduce)
```
Missing `weight` hook.

sneaxiy commented on fleetx/models/gpt_model/modeling_hybrid.py:
```diff
@@ -912,464 +1014,3 @@ def from_config(cls, cfg):
             "fused_linear": cfg['fused_linear'],
             "recompute_granularity": cfg['recompute_granularity']
         }
-
```
Why remove this?

sneaxiy commented on fleetx/models/gpt_model/modeling_hybrid.py:
```diff
@@ -99,63 +104,88 @@ def __init__(self,
+                fuse_matmul_bias=fused_linear,
+                sequence_parallel=sequence_parallel)
 
             self.k_proj = fleet.meta_parallel.ColumnParallelLinear(
```
`ColumnParallelLinear = XXXColumnParallelLinear if sequence_parallel else fleet.meta_parallel.ColumnParallelLinear`
`RowParallelLinear = XXXRowParallelLinear if sequence_parallel else fleet.meta_parallel.RowParallelLinear`

GhostScreaming committed:
Merge remote-tracking branch 'upstream/develop' into sequence_parallel

GhostScreaming committed:
Fix merge conflict of modeling_hybrid.py

GhostScreaming committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/FleetX into sequence_parallel

GhostScreaming committed:
Sequence parallel has been tested. When mp parallelism is 2,
percision of each layer is aligned in the first forward pass
and backward pass, and loss curve matched with the peer after
6000 steps.

GhostScreaming committed:
Add config file for sequence parallel (mp=8)

GhostScreaming committed:
Remove useless print sentence, and run.sh script.

GhostScreaming committed:
Add sequence_parallel option in gpt_config.py

sneaxiy commented on examples/gpt/hybrid_parallel/configs_1.3B_dp8.yaml:
```diff
@@ -51,13 +51,13 @@ Model:
     sharding_degree: 1
     sharding_stage: 1
-    sharding_offload: False
+    sharding_offload: True
```
Why change this?

sneaxiy commented on examples/gpt/hybrid_parallel/configs_1.3B_mp8.yaml:
```diff
@@ -0,0 +1,84 @@
+# 1.3B
```
Do we really need this file?

sneaxiy commented on examples/gpt/hybrid_parallel/run.sh:
```diff
@@ -1,24 +0,0 @@
-# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Why remove this file?

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+    group.process_group.allreduce(grad).wait()
+    return grad
+
+# def gradient_hook(grad, name="gradient", prefix="seq_paral", is_break=True):
```
Remove useless code.

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+                 name=None):
+        super(ColumnSequenceParallelLinear, self).__init__()
+
+        self.model_parallel_group = tp._HYBRID_PARALLEL_GROUP.get_model_parallel_group(
```
Replace `tp._HYBRID_PARALLEL_GROUP` with `fleet.get_hybrid_communicate_group()`. Same below.

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
I prefer to renaming this file to be `sequence_parallel_utils.py`.

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+# All gather along the first dim during forward pass
+# All reduce and scatter along the first dim during backward pass
+class AllGatherOp(PyLayer):
+    # input shape: [s/n, b, h], n is mp parallelism
```
Duplicate code with `ReduceScatterOp`. Can use two functions to simplify the codes.

```Python
def all_gather(...): pass

def reduce_scatter(...): pass

class ReduceScatterOp(PyLayer):
    @staticmethod
    def forward(ctx, ...):
        return reduce_scatter(...)

    @staticmethod
    def backward(ctx, ...):
         return all_gather(...)

class AllGatherOp(PyLayer): ...
```

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+    # input shape: [s, b, h], n is mp parallelism
+    # after forward shape: [s/n, b, h]
+    @staticmethod
+    def forward(self, input):
```
The first argument of `PyLayer.forward` is `ctx` but not `self`. Although there is no runtime error, someone would think the word `self` means that this method belongs to an object but not the class.

Same below.

sneaxiy commented on fleetx/models/gpt_model/modeling_hybrid.py:
```diff
@@ -915,15 +985,12 @@ def from_config(cls, cfg):
             "recompute_granularity": cfg['recompute_granularity']
         }
 
-
```
Why remove these empty lines?

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+        group.process_group.all_gather(grad, output).wait()
+        return output
+
+class GatherOp(PyLayer):
```
Duplicate codes with `ScatterOp`. Please follow the code suggestions in `ReduceScatterOp`.

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+        self._dtype = self._helper.get_default_dtype()
+        self._name = name
+
+        self.model_parallel_group = tp._HYBRID_PARALLEL_GROUP.get_model_parallel_group(
```
Same above. Replace `tp._HYBRID_PARALLEL_GROUP`.

sneaxiy commented on fleetx/models/gpt_model/utils.py:
```diff
@@ -0,0 +1,369 @@
+
+        self.weight.is_distributed = True if self.is_mp else False
+
+        # if sequence parallel is true, 
```
Remove useless codes.

GhostScreaming committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/FleetX into sequence_parallel

sneaxiy commented on examples/gpt/hybrid_parallel/configs_1.3B_dp8.yaml:
```diff
@@ -48,10 +48,9 @@ Model:
   use_recompute: True
   recompute_granularity:
   fused_linear: True
-  
```
This file is not necessary to be changed.

sneaxiy commented on fleetx/models/gpt_model/sequence_parallel_utils.py:
```diff
@@ -0,0 +1,342 @@
+        self._dtype = self._helper.get_default_dtype()
+        self._name = name
+
+        # self.model_parallel_group = fleet.get_hybrid_communicate_group(
```
Remove useless codes.

sneaxiy commented on fleetx/models/gpt_model/sequence_parallel_utils.py:
```diff
@@ -0,0 +1,342 @@
+            input_parallel = AllGatherOp.apply(x)
+        else:
+            input_parallel = x
+        output = self.linear(input_parallel,
```
Try to run `pre-commit`.

sneaxiy commented on fleetx/models/gpt_model/sequence_parallel_utils.py:
```diff
@@ -0,0 +1,342 @@
+            self.linear = fused_linear
+
+    def forward(self, x):
+        if self.input_is_parallel or (not self.is_mp):
```
The `self.input_is_parallel` is always True. Try to remove the `else` codes.

GhostScreaming committed:
Fix some code style problems.
1. Remove configs_1.3B_mp8.yaml
2. Add run.sh
3. Rename utils.py to sequence_parallel_utils.py
4. Fix some code style problems in modeling_hybrid.py
5. If sequence parallel is True, encoder_output should  all_gather and
       transpose

sneaxiy reviewed (approved):
LGTM.

GhostScreaming committed:
Withdraw modification of run.sh

GhostScreaming committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/FleetX into sequence_parallel

GhostScreaming committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/FleetX into sequence_parallel
